aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2546,13 +2569,22 @@ Tensor mexp_impl(
 ) {
   auto res = at::empty_like(a);
   const auto norm = operator_1_norm(a);
-  // `norm_cpu` is used to decide which Tensors require which approximation
-  // based on their norm. This decision takes place on CPU.
-  // It requires moving data back and forth between devices when `a` is on CUDA,
-  // but at the cost of only one sigle CPU-CUDA synchronization (instead of 6),
+  const auto norm_max = norm.max().to(at::kCPU).item().toFloat();
+  if (std::isnan(norm_max)) {
+    res.fill_(std::numeric_limits<double>::quiet_NaN());
+    return res;
+  }
+
+  // `norm_small_to_cpu` is used to decide which Tensors require which
+  // approximation based on their norm. This decision may take place on CPU.
+  // It could require moving data back and forth between devices when `a` is on CUDA,
+  // but at the cost of only one single CPU-CUDA synchronization (instead of 6),
   // and better performance overall (benchmarked).
-  const auto norm_cpu = (a.device().type() == at::kCUDA)
-    ? norm.to(at::kCPU) : norm;
+  // For large batch size, the cost of the above decision on CPU will be quite 
+  // expensive (compared to the synchronization overhead from GPU to CPU), so
+  // here we have a threshold to determine whether or not to move norm to CPU.
+  const auto norm_small_to_cpu = ((a.device().type() == at::kCUDA) 
+    && a.size(0) >= large_batch_threshold) ? norm.to(at::kCPU) : norm;
```

Comment:
Here you are sending large batches to CPU, which is the opposite to what you claim you are doing. Please provide benchmarks that back up this optimisation.